design/general-progress-monitoring.md

```diff
@@ -0,0 +1,462 @@
+plugins.  While the majority of BackupItemAction plugins do not take snapshots or upload data, this
+functionality is useful for any longstanding plugin operation managed by an external
+process/controller so we will modify BackupItemAction and RestoreItemAction to optionally return an
+operationID in addition to the modified item.
+
+The Backup/RestoreItemAction plugin identifier as well as the Item and Snapshot ID will be stored in
+the `<backup-name>-itemoperations.json.gz`.  When checking for progress, this info will be used to
+select the appropriate Backup/RestoreItemAction plugin to query for progress.
+
+Velero can attempt to cancel an operation by calling the Cancel API call on the BIA/RIA. The plugin
+can then take any appropriate action as needed. Cancel will be called on backup deletion, and
+possibly upon reaching timeouts, if timeout support is included in this feature.
+
+_NotApplicable_ should only be returned if the Backup/RestoreItemAction plugin should not be
+handling the item.  If the SnapshotItemAction plugin should handle the item but, for example, the
+item/snapshot ID cannot be found to report progress, an OperationProgress struct with the error set
+appropriately (in this case _NotFound_) should be returned. If the item action does not start an
+asynchronous operation, then operationID will be empty.
+
+Two new methods will be added to the BackupItemAction interface, and the Execute() return signature
+will be modified:
+
+    	// Execute allows the ItemAction to perform arbitrary logic with the item being backed up,
+    	// including mutating the item itself prior to backup. The item (unmodified or modified)
+    	// should be returned, an optional operationID, along with an optional slice of ResourceIdentifiers
+    	// specifying additional related items that should be backed up. If operationID is specified
+    	// then velero will wait for this operation to complete before the backup is marked Completed.
+    	Execute(item runtime.Unstructured, backup *api.Backup) (runtime.Unstructured, operationID string,
+    	    []ResourceIdentifier, error)
+    	
+    	// Progress  
+    	Progress(input *BackupItemActionProgressInput) (OperationProgress, error)
+    	// Cancel  
+    	Cancel(input *BackupItemActionProgressInput) error
+    
+    // BackupItemActionProgressInput contains the input parameters for the BackupItemAction's Progress function.
+    type BackupItemActionProgressInput struct {
+    	// Item is the item that was stored in the backup
+    	Item runtime.Unstructured
+    	// OperationID is the operation ID returned by BackupItemAction Execute
+    	operationID string
+    	// Backup is the representation of the backup resource processed by Velero.
+    	Backup *velerov1api.Backup
+    }
+
+Two new methods will be added to the RestoreItemAction interface, and the
+RestoreItemActionExecuteOutput struct will be modified:
```

Comment:
RestoreItemActionExecuteOutput should be changed to RestoreItemAction interface.

Reply:
RestoreItemActionExecuteOutput is the return type of the Execute function in the RestoreItemAction interface. We need to add a new field to this struct, so the above is correct.